### DISCUSSION TITLE
EditLevelMismach exceptions after save

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#3580] [Questions]
We've hit something odd while testing our app using http proxy after switching from wcf.  I'm posting b/c we've hit this same issue on multiple forms/BOs, and while I'm sure it could be something in our code, its just unusual (and if we go back to WCF the exceptions go away).

The issue seems to be binding the object returned from calling Save on the BO; as soon as we set the BindingSource.DataSource to the newly returned object, we get an edit level mismatch exception.  Fetching an object is fine, its just after the roundtrip saving.

I'm at a bit of a loss, b/c for the object graph i'm looking at, none of the BOs needed changes to work over httpproxy.  They're pretty normal BOs.

As a bit of a wrinkle, if I enumerate all the child objects the root object has, they all have EditLevel 0, as does the root.  So I'm even more confused.

.Net framework 4.7.2, Csla 5.4.2

**Comment by rockfordlhotka** [comment]
Switching from WCF to Http also means that you switched from NDCS (basically BinaryFormatter) to MobileFormatter.

It is possible that some field in `BusinessBase` isn't being properly serialized via MobileFormatter - or something like that. Though the fact that you've confirmed that EditLevel is 0 for all the objects kind of flies in the face of this theory.

Still, it is almost certain that some value must not be flowing through serialization as expected.

Are you saying that some graphs work and some don't? Can you identify any type in particular that is used only in the failing graphs? It might not be `BusinessBase` - maybe it is one of the list types or something.

**Reply by ajohnstone-ks** [reply]
Yes, I'm actually pretty sure it must be something in our own code, its just that I can't see it.  It is only some BOs, not all of them.  But it was not something I was expecting, and given the number that have popped up I just thought I'd ask here to see if someone had something similar.

Regarding the CopyState; does that only look at managed fields when doing the copy, or is it also reflecting over private things?  We do a base logging class (which keeps a Dictionary<string, object> that we manually serialize) for access to the values as the object was loaded.  I thought there's a not-undoable on it, but maybe not, or maybe there's something else similar in the graph that should be ignored.

So I did verify there's **[field: NotUndoable]** on the property (it's an auto property), but that should keep it ignored by CopyState correct?

**Reply by rockfordlhotka** [reply]
One of the limitations of MobileFormatter is that it doesn't honor `NotUndoable` (I don't think). I _know_ it doesn't honor `NonSerialized`.

To make something not serialize through MobileFormatter, you need to make the property be not managed (use a private backing field) and then do not serialize/deserialize that backing field in the `OnGetState`/`OnSetState` method overrides.

**Reply by rockfordlhotka** [reply]
A corollary to this is that MobileFormatter does not serialize private fields. It only automatically serializes managed properties, and you can opt-in to serializing private fields by overriding the get/set state methods.

**Reply by ajohnstone-ks** [reply]
Ok, this property I'm talking is already field/non-managed property.  Its contents though we serialize ourselves and that does seem to work fine.

I've confirmed EditLevels are all 0 once we're back on the client, its when we try to rebind the new BO to the form it blows up right away.  Oddly, its only after the object has been saved as well.  The first time opening the form when we Fetch from the server, its fine.  Just after successfully returning to the client and during our rebinding, something throws this.  I can actually fix it by fetching new from the server again instead of using the obj returned from Save.  Which is part of the reason for my confusion on this.

**Reply by rockfordlhotka** [reply]
Are you using the save and merge feature, or doing the traditional replacement of the bound object?

**Reply by ajohnstone-ks** [reply]
Traditional replacement of bound object.
